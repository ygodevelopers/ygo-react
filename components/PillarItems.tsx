import { Pillar } from '@/types';
import { useEffect, useState } from "react";
import { Text, TouchableOpacity, View, Image, TextInput,StyleSheet, Dimensions,Modal } from "react-native";
import { Router } from "expo-router";

import PillarAddColor from '@/components/PillarAddColor';
import PillarAddIcon from '@/components/PillarAddIcon';
import {usePillar} from '@/context/pillarContext';
import 'react-native-get-random-values';
import { v4 as uuidv4 } from "uuid";



const screenWidth = Dimensions.get('window').width;
const ITEM_WIDTH = (screenWidth - 16 * 2 - 16) / 2; 


export const PillarItems=({item, router}:{item: Pillar, router: Router})=> {
    
    const {         
        selectedColor, 
        pillarname, 
        selectedicon,
        getPillars,
        savePillars,
        modalNewVisible,
        setModalNewVisible,
        subpillar,
        setsubpillar,
        addSubPillar,
        currentPillar,
        setcurrentPillar
        } = usePillar();

        const emojis = Array.from(item.icon || "");
        const emoji1 = emojis[0];
        const emoji2 = emojis[1];

    const handlePress = () => {

        if (item?.title === 'New Pillar') {
            setModalNewVisible(true);
        }
        else {

             router.push({pathname: '/pillarDetail', params: {pillarId: item.id, pillarTitle: item.title, pillarIcon:item.icon,subpillars: JSON.stringify(item.subPillars) }});
             setcurrentPillar(item);
            // item={item} setModalEditVisible={setModalEditVisible} handleSubPillar={handleSubPillar}
    
           // setModalEditVisible(true);
        }
        // Add your navigation or action here
    };

    const handleSavePillar = async() => {
        if(subpillar){
            const newSub: Pillar = {
            //get uuid with capital letter
            id: uuidv4().toUpperCase(),
            title: pillarname,
            icon: selectedicon,
            color: selectedColor,
            type: 'sub',
            subPillars: [],
            userId: currentPillar?.userId, // Assuming you have a userId in the item
            };
            const result = await addSubPillar(currentPillar.id, newSub);
            if (result.success) {
                let item = currentPillar;
                item.subPillars.push(newSub);
                setcurrentPillar(item);
            } else {
                console.error("SubPillar Failed:", result.msg);
            }
            setsubpillar(false);

        }else{
            await savePillars(uuidv4().toUpperCase(), pillarname, selectedColor, selectedicon);     
        }
        getPillars();
        setModalNewVisible(false);
    };

    return (
        <>
            <TouchableOpacity onPress={handlePress} activeOpacity={0.7}>
                <View style={{
                    flexDirection: 'row',
                    padding: 10, 
                    marginHorizontal: 10,
                    marginVertical: 5, 
                    width: ITEM_WIDTH,
                    borderWidth: 2,
                    borderColor: '#aaa',
                    borderRadius: 12,
                    gap: 8,
                    elevation: 25,
                    backgroundColor: 'white'
                    }} >
                        {item.type === 'new' ? 
                        (
                            <Text style={{ fontSize: 24, width:40, color:'red'}}>{item?.icon}</Text>
                        )
                        
                        :(
                            <>
                            <Text style={[styles.icon,{borderColor: item.color}]}>{emoji1}</Text>
                            {/* <Text style={{ fontSize: 24, width:40, borderWidth: 3, borderColor: item.color, borderRadius: 10, textAlign: 'center',textAlignVertical: 'center'}}>{emoji2}</Text> */}
                            </>
                        )}
                        <Text style={styles.title}>{item?.title}</Text>
                </View>
            </TouchableOpacity>
            <Modal visible={modalNewVisible} animationType="slide">
                <View style={{ flex: 1, alignItems: 'center' }}>
                    <View style={{ flexDirection: 'row', justifyContent: 'space-between', width: '100%', paddingHorizontal: 20, marginBottom: 20,marginTop: 20  }}>
                        <TouchableOpacity onPress={() => setModalNewVisible(false)}>    
                            <Text>Cancel</Text>
                        </TouchableOpacity>
                        <Text>Create Pillar</Text>
                        <TouchableOpacity onPress={() => handleSavePillar()}>    
                            <Text>Save</Text>
                        </TouchableOpacity>
                    </View>
                    <NewPillar />
                </View>
            </Modal>
        </>
    );
}

function NewPillar(){
    const {selectedColor, pillarname, setpillarName, selectedicon, setselectedicon} = usePillar();
    useEffect(() => {
        setselectedicon('ðŸ“¸');
    }, []);
    return(
        <View style={{ flexDirection: 'column', alignItems: 'center' }}>

            <Text style={{ fontSize: 48,  marginRight: 0, borderWidth: 3, borderColor: selectedColor, borderRadius: 15 ,color: 'gold'}}>{selectedicon}</Text>
            <View style={{ flexDirection: 'row'}}>
                <PillarAddIcon />
                <PillarAddColor />
            </View> 
            <TextInput
                    style={{borderWidth: 1,
                            borderColor: '#ccc',
                            padding: 10,
                            fontSize: 16,
                            borderRadius: 6,
                            marginBottom: 20,}}
                    placeholder="Pillar Name"
                    value={pillarname}
                    onChangeText={setpillarName}
                />
        </View>
   
    )
}

const styles = StyleSheet.create({
  container: { flex: 1, justifyContent: 'center', alignItems: 'center' },
  icon: {  fontSize: 24, width:40, borderWidth: 3, borderRadius: 10, textAlign: 'center',textAlignVertical: 'center' },
  title: { fontSize: 18, marginTop: 8 },
});