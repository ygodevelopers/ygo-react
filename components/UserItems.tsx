import { Text, TouchableOpacity, View, Image } from "react-native";
import { useEffect, useState } from "react";
import { Pillar, Thread, User } from "@/types";
import { Router } from "expo-router";
import { useAuth } from "@/context/authContext";
import { usePillar } from "@/context/pillarContext";
import { userRef } from "@/firebaseConfig";
import { getDoc, doc } from "firebase/firestore";
import tinycolor from "tinycolor2";


export default function UserItems({ item, router }: { item: Thread, router: Router }) {

    const { Pillars, setCurrentThread } = usePillar();
    const { user } = useAuth();
    const [contact, setContact] = useState<User>();
    const [pillar, setPillar] = useState<Pillar>();


    useEffect(() => {
        getUserInfo();
        getPillarInfo();
        setCurrentThread(item);
    }, [item]);



    if (!user) return null;
    const openChatRoom = () => {
        const threadID = item.id;
        router.push({ pathname: '/(app)/chatRoom', params: { threadID: threadID, contactName: contact?.firstName, contactID: contact?.id } });
    }

    const getUserInfo = async () => {
        const { uids } = item;
        const otherUserId = uids.find((uid) => uid !== user?.id);
        if (!otherUserId) return;

        try {
            const userDoc = await getDoc(doc(userRef, otherUserId));
            if (userDoc.exists()) {
                setContact(userDoc.data() as User);
            }
        } catch (error) {
            console.error("Failed to fetch contact info:", error);
        }
    };

    const getPillarInfo = () => {
        Pillars.forEach((element: Pillar) => {
            item.pillarId?.forEach(threadPillar => {
                if (element.id == threadPillar) {
                    setPillar(element);
                }
            })
        });
    }

    // Format time like iOS Messages, make it state based to update when minute changes? 
    const formatTime = () => {
        if (!item.lastUpdated) return "";

        const messageDate = item.lastUpdated.toDate();
        const now = new Date();
        const diffInMinutes = Math.floor((now.getTime() - messageDate.getTime()) / (1000 * 60));
        const diffInHours = Math.floor(diffInMinutes / 60);
        const diffInDays = Math.floor(diffInHours / 24);

        if (diffInMinutes < 60) {
            return `${diffInMinutes}m`;
        } else if (diffInHours < 24) {
            return `${diffInHours}h`;
        } else if (diffInDays < 7) {
            return `${diffInDays}d`;
        } else {
            return messageDate.toLocaleDateString('en-US', {
                month: 'short',
                day: 'numeric'
            });
        }
    }

    return (
        <TouchableOpacity onPress={openChatRoom} className={"flex-row items-center px-4 py-3 bg-white active:bg-gray-50"}>
            <View className="mr-4 items-center justify-center" style={{
                borderRadius: 100,
                borderWidth: 4,
                borderColor: pillar?.color || "#d1d5db",
            }}>
                {contact?.profileImageUrl ? (
                    <Image
                        source={{ uri: contact.profileImageUrl }} className="w-12 h-12 rounded-full" style={{ width: 62, height: 62 }}
                    />
                ) : (
                    <View className="w-12 h-12 rounded-full bg-gray-300 items-center justify-center">
                        <Text className="text-white font-semibold text-lg">
                            {`${contact?.firstName?.charAt(0) || ""}${contact?.lastName?.charAt(0) || ""}`.toUpperCase() || "?"}
                        </Text>
                    </View>
                )}
            </View>

            <View className="flex-1 border-b border-gray-200 pb-3">
                <View className="flex-row items-center justify-between mb-1">
                    <Text className="text-black font-semibold text-lg">{contact?.firstName || "Unknown"} {contact?.lastName || "Unknown"}</Text>
                    <Text className="text-gray-500 text-sm"> {formatTime()}</Text>
                </View>
                {
                    pillar && (
                        <View style={{borderRadius: 10, backgroundColor: pillar.color, alignSelf: 'flex-start'}} className="justify-center items-center px-2">
                            <Text className="text-sm font-medium mb-1" style={{ color: tinycolor(pillar.color).darken(40).toString()}}>{pillar.title}</Text>
                        </View>
                        
                )}
                <Text className="text-gray-500 text-base" numberOfLines={1}>
                    {(() => {
                        if (!item.lastMessage?.messageText) return "";
                        
                        const messageText = item.lastMessage.messageText;
                        const isFromUser = item.lastMessage.fromId === user.id;
                        
                        const prefix = isFromUser ? "You: " : "";
                        return `${prefix}${messageText.slice(0, 7)}...`;
                    })()}
                </Text>
            </View >
        </TouchableOpacity >
    );
}
