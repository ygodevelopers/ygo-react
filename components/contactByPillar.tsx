import { Contact,User } from "@/types";
import { useEffect, useState } from 'react';
import { View, FlatList, Image, Text, StyleSheet,TouchableOpacity } from 'react-native';
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';
import { getDocs, setDoc, doc, query, where, updateDoc ,arrayUnion } from 'firebase/firestore'
import { contactCollection, userRef } from '@/firebaseConfig';
import {AlphabetList, IData} from 'react-native-section-alphabet-list';
import { useAuth } from '@/context/authContext';
import {usePillar} from '@/context/pillarContext';
import { Router, useRouter } from "expo-router";

export default function ContactByPillar(){

  const router = useRouter();

  const {user} = useAuth();

  const {
      currentPillar,
      currentThread
    } = usePillar();

  const [userContacts, setUserContacts] = useState<User[]>();
  useEffect(() => {
        getContacts();
    }, []);
  
  const getContacts = async () => {
    try {
        const contactsContainer : Contact[] = [];
        const q = query(contactCollection, where('ownerId', '==', user.id),where('pillarId', 'array-contains', currentPillar?.id));
        const qSnapshot = await getDocs(q);
        qSnapshot.forEach((doc) => {
          contactsContainer.push(doc.data() as Contact)
        });
        getUsersFromContacts(contactsContainer);
    } catch (error) {
        console.error("Error fetching contact:", error);
    }
  }

  const getUsersFromContacts = async (contacts: Contact[]) => {
      const promises = contacts.map(async (contact) => {
          const q = query(userRef, where('id', '==', contact.contactUserId));
          const qSnapshot = await getDocs(q);
          const user : User[] = [];
          qSnapshot.forEach((doc) => {
              user.push(doc.data() as User);
          })
          return user;
      });

      const result = await Promise.all(promises);
      const usersContainer : User[] = result.flat();
      setUserContacts(usersContainer);
  }
return (
    <View style={styles.container} >
    {
          userContacts && 
          <FlatList
            data = {userContacts}
            contentContainerStyle = {{flex:1, paddingVertical: 55}}
            keyExtractor={(item, index) => index.toString()}
            showsVerticalScrollIndicator = {false}
            renderItem={({item, index})=>(          
                // <TouchableOpacity onPress={() => router.push({ params: { contactID: item.id, threadID: null }, pathname: '/(app)/ContactView' })}>
                <TouchableOpacity onPress={() => router.push({ pathname: '/(app)/chatRoom', params: { threadID: currentThread?.id, contactName: item?.firstName, contactID: item?.id } })}>
                 
                  <View style={{
                      flexDirection: 'row',
                      padding: 10, 
                      marginHorizontal: 10,
                      marginVertical: 5, 
                      width: 300,

                      gap: 8,}} >

                          <Image source={item?.profileImageUrl ? { uri: item.profileImageUrl } : undefined} style={{borderRadius: 100, height: hp(4.5), aspectRatio: 1}}/>

                          <Text>{item?.firstName} {item?.lastName}</Text>
                  </View>
               </TouchableOpacity>
              )}
            />
      }
    </View>
  );
}
const styles = StyleSheet.create({
  container: {
              flexDirection: 'row',
              padding: 10, 
              marginHorizontal: 10,
              marginVertical: 5, 
              width: 350,

              gap: 8},
  icon: { fontSize: 24 },
  title: { fontSize: 24 },
});