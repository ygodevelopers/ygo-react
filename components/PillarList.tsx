import { ActivityIndicator, Text, View, FlatList, StyleSheet, TouchableOpacity} from "react-native";
import { useEffect, useState } from "react";

import { StatusBar } from 'expo-status-bar'
import { useAuth } from '@/context/authContext'
import { PillarItems } from '@/components/PillarItems'
import {usePillar} from '@/context/pillarContext'
import { useRouter} from "expo-router";



// type Props = NativeStackScreenProps<PillarStackParamList, 'Home'>;

export const PillarList = () => {
    const {user} = useAuth();
    const {getPillars, Pillars, loading} = usePillar();

    const router = useRouter();

    useEffect(()=>{
          if(user?.id){
            getPillars();
        }
    },[user])


    const combinedData = [...Pillars, { icon: 'âž•', title: 'New Pillar',type:'new' }];

    return (
        <View
            style={{
                flex: 1,
                // justifyContent: "center",
                // alignItems: "center",
            }}
            >
            <StatusBar style="light" />      
            {/* <Text>Pillars List</Text>       */}
            {
                !loading ? (
                    <View className = "flex-1">
                        <FlatList
                            data = {combinedData}
                            contentContainerStyle = {{flex:1, paddingVertical: 55}}
                            keyExtractor={(item, index) => index.toString()}
                            numColumns={2}
                            showsVerticalScrollIndicator = {false}
                            columnWrapperStyle={{justifyContent: 'space-between'}}
                            renderItem={({item, index})=><PillarItems item={item} router={router}/>}
                            />
      
                    </View>
                ):(
                    <View className="flex item-center">
                        <ActivityIndicator size="large" />
                    </View>
                )
            }
        </View>
    )
}
