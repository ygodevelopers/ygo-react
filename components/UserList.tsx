import { View, FlatList, Text } from "react-native";
import { useEffect, useState } from "react";
import { useAuth } from '@/context/authContext'
import { StatusBar } from 'expo-status-bar'
import UserItems from '@/components/UserItems'
import { Thread } from "@/types";
import { useRouter } from "expo-router";
import { subscribeToThreads } from "@/utils/chatService";
import { onSnapshot } from "firebase/firestore";
import { userRef } from "@/firebaseConfig";
import { doc } from "firebase/firestore";

export const UserList = ({ pillarid }: { pillarid?: string | null }) => {
    const { user } = useAuth();
    const [threads, setThreads] = useState<Thread[]>([]);
    const [blockedUserIds, setBlockedUserIds] = useState<string[]>([]);
    const router = useRouter();
    const pid = pillarid ?? null;
    

    useEffect(() => {
        if (!user?.id) return;
        const unsubscribe = subscribeToThreads(user.id, (threadsArray) => {
            const filteredThreads = threadsArray.filter(thread => {
                const otherUserId = thread.uids.find((uid: string) => uid !== user.id);
                const isUserAllowed = !blockedUserIds.includes(otherUserId || "");
       
                const isPillarMatched = pid ? thread.pillarId?.includes(pid) : true;
                return isUserAllowed && isPillarMatched;
            });
            filteredThreads.sort(sortThreads);
            setThreads(filteredThreads);
        });

        return () => {
            if (unsubscribe) {
                unsubscribe();
            }
        };
    }, [user?.id, blockedUserIds]);


    useEffect(() => {
        if (!user?.id) return;

        const unsubscribe = onSnapshot(doc(userRef, user.id), (docSnapshot) => {
            const data = docSnapshot.data();
            setBlockedUserIds([...(data?.blockedUsers || [])]);
        });

        return () => unsubscribe();
    }, [user?.id]);

    const sortThreads = (threadOne: Thread, threadTwo: Thread) => {

        const timeOne = threadOne.lastUpdated?.seconds ?? 0;
        const timeTwo = threadTwo.lastUpdated?.seconds ?? 0;

        if (timeOne < timeTwo) {
            return 1;
        }
        if (timeOne > timeTwo) {
            return -1;
        }
        return 0;
    }

    return (
        <View className="flex-1 bg-white">
            <StatusBar style="dark" />
            {
                threads.length > 0 ? (
                    <View className="flex-1">
                        <FlatList
                            data={threads}
                            contentContainerStyle={{ paddingVertical: 25 }}
                            keyExtractor={(item: Thread, index) => index.toString()}
                            showsVerticalScrollIndicator={false}
                            renderItem={({ item, index }) => item.lastUpdated ? (<UserItems item={item} router={router} />) : null}
                            className="flex-1"
                        />
                    </View>
                ) : (
                    <View className="flex-1 items-center justify-center">
                        <Text className="text-gray-500 text-lg">No Messages found</Text>
                    </View>
                )
            }
        </View>
    )
}