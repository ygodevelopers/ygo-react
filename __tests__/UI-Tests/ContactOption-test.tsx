import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { FontAwesome6 } from '@expo/vector-icons';
import { ContactOption } from '@/components/ContactOption';

// Mock the FontAwesome6 component
jest.mock('@expo/vector-icons', () => ({
  FontAwesome6: ({ name, size, color, testID, ...props }) => {
    const MockedIcon = require('react-native').Text;
    return (
      <MockedIcon testID={testID || 'mocked-icon'} {...props}>
        {name}
      </MockedIcon>
    );
  },
}));

describe('ContactOption Component', () => {
    const mockHandlePress = jest.fn();
    const mockSymbol = <FontAwesome6 name="phone" size={20} color="#374151" testID="contact-icon" />;
    const mockText = 'Call Support';

    beforeEach(() => {
        mockHandlePress.mockClear();
    });

    describe('Rendering', () => {
        it('renders correctly with all props', () => {
            const { getByTestId, getByText } = render(
                <ContactOption
                symbol={mockSymbol}
                text={mockText}
                handlePress={mockHandlePress}
                />
            );
            expect(getByText(mockText)).toBeTruthy();
            expect(getByTestId('contact-icon')).toBeTruthy();
        });

        it('renders the provided text correctly', () => {
            const customText = 'Send Email';
            const { getByText } = render(
                <ContactOption
                symbol={mockSymbol}
                text={customText}
                handlePress={mockHandlePress}
                />
            );
            expect(getByText(customText)).toBeTruthy();
        });

        it('renders the provided symbol component', () => {
            const customSymbol = <FontAwesome6 name="email" size={20} color="#374151" testID="email-icon" />;
            const { getByTestId } = render(
                <ContactOption
                symbol={customSymbol}
                text={mockText}
                handlePress={mockHandlePress}
                />
            );
            expect(getByTestId('email-icon')).toBeTruthy();
        });

        it('matches snapshot', () => {
            const component = render(
                <ContactOption
                symbol={mockSymbol}
                text={mockText}
                handlePress={mockHandlePress}
                />
            );

            expect(component.toJSON()).toMatchSnapshot();
        });
    });

    describe('Props validation', () => {
        it('handles different text lengths correctly', () => {
            const longText = 'This is a very long contact option text that should still render properly';
            const { getByText } = render(
                <ContactOption
                symbol={mockSymbol}
                text={longText}
                handlePress={mockHandlePress}
                />
            );
            expect(getByText(longText)).toBeTruthy();
        });

        it('handles empty text', () => {
            const { queryByText } = render(
                <ContactOption
                symbol={mockSymbol}
                text=""
                handlePress={mockHandlePress}
                />
            );
            expect(queryByText('')).toBeTruthy();
        });

        it('handles different icon components', () => {
            const customIcon = <FontAwesome6 name="chat" size={24} color="#059669" testID="chat-icon" />;
            const { getByTestId } = render(
                <ContactOption
                symbol={customIcon}
                text={mockText}
                handlePress={mockHandlePress}
                />
            );

            expect(getByTestId('chat-icon')).toBeTruthy();
        });
    });

    describe('Multiple instances', () => {
        it('renders multiple ContactOption components correctly', () => {
            const options = [
                { symbol: <FontAwesome6 name="phone" testID="phone-icon" />, text: 'Call', handler: jest.fn() },
                { symbol: <FontAwesome6 name="email" testID="email-icon" />, text: 'Email', handler: jest.fn() },
                { symbol: <FontAwesome6 name="chat" testID="chat-icon" />, text: 'Chat', handler: jest.fn() }
            ];

            const { getByText, getByTestId } = render(
                <>
                {options.map((option, index) => (
                    <ContactOption
                    key={index}
                    symbol={option.symbol}
                    text={option.text}
                    handlePress={option.handler}
                    />
                ))}
                </>
            );

            options.forEach((option) => {
                expect(getByText(option.text)).toBeTruthy();
            });

            expect(getByTestId('phone-icon')).toBeTruthy();
            expect(getByTestId('email-icon')).toBeTruthy();
            expect(getByTestId('chat-icon')).toBeTruthy();
        });
    })
});