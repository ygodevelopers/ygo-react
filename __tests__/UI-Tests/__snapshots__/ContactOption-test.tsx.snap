// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactOption Component Rendering matches snapshot 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    className="flex-row justify-start items-center gap-4 py-2 px-1"
  >
    <View
      className="w-6 items-center"
    >
      <Text
        testID="contact-icon"
      >
        phone
      </Text>
    </View>
    <Text
      className="text-base text-gray-800 flex-1"
    >
      Call Support
    </Text>
  </View>
</View>
`;
